AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/MutationSync/OutgoingMutationQueue/SyncMutationToCloudOperation.swift

```diff
@@ -139,7 +172,36 @@ class SyncMutationToCloudOperation: Operation {
                 self.finish(result: asyncEvent)
             }
         }
+    }
 
+    private func getRetryAdviceIfRetryable(error: APIError) -> RequestRetryAdvice {
+        var advice = RequestRetryAdvice(shouldRetry: false, retryInterval: DispatchTimeInterval.never)
+
+        switch error {
+        case .networkError(_, _, let error):
+            //currently expecting APIOperationResponse to be an URLError
+            let urlError = error as? URLError
+            advice = requestRetryablePolicy.retryRequestAdvice(urlError: urlError,
+                                                               httpURLResponse: nil,
+                                                               attemptNumber: currentAttemptNumber)
+        case .httpStatusError(_, let httpURLResponse):
+            advice = requestRetryablePolicy.retryRequestAdvice(urlError: nil,
+                                                               httpURLResponse: httpURLResponse,
+                                                               attemptNumber: currentAttemptNumber)
+        default:
+            break
+        }
+        return advice
+    }
+
+    private func scheduleRetry(advice: RequestRetryAdvice) {
+        log.verbose("\(#function) scheduling retry for mutation")
+        mutationRetryNotifier = MutationRetryNotifier(retryAttemptNumber: currentAttemptNumber,
+                                                      advice: advice) {
+                                                        self.sendMutationToCloud()
+                                                        self.mutationRetryNotifier = nil
+        }
+        currentAttemptNumber += 1
```

Comment:
I know these mutation operations are enqueued on a serial Operations queue; is any of the internal work being done in a way that makes this not-thread-safe? My quick scan makes me think everything is pretty well serialized, but wanted you to confirm since you're closer to this code.

Reply:
Just to double check, your question revolves around executing multiple SyncMutationToCloudOperations running at the same time such that we could run into a situation where our code is not thread safe?  (not that we'd ever want to do this, because we want to drain the mutation queue one at a time in case any of the data changes)

But... I looked at the code and thought about implications of instance variables if we were to create multiple operations and run them at the same time, and didn't see a situation where we would be overwriting data.

Reply:
No, I'm actually concerned about multiple operations happening on a single operation--e.g., a network callback being invoked at the same time as we're trying to make the outgoing call b/c of a retry.

Reply:
I see... hmmm.. I suppose it is possible that our timer could wake up at the same exact time that we have a receive() through combine (e.g. a network event).

In this case, I think a pragmatic solution could be to introduce something (atomic write to a variable?) in our `notifyCallback` function which allows the "first event" to win.